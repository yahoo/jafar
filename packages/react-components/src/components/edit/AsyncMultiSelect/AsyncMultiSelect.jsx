/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import SelectInternal from 'react-select';
import { customStyles } from '../Select/Select';

/**
 * Represent a selection of multi items from async items list 
 * 
 * Import <a target="_blank" href=
 "https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/AsyncMultiSelect/AsyncMultiSelect.jsx">
 AsyncMultiSelect</a>  from '@jafar/react-components/edit/AsyncMultiSelect'
 */
export default class AsyncMultiSelect extends React.Component {
  static propTypes = {
    value: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.any,
    })),
    state: PropTypes.shape({
      items: PropTypes.arrayOf(PropTypes.shape({
        label: PropTypes.string,
        value: PropTypes.any,
      })),
      placeholder: PropTypes.string,
      searchQuery: PropTypes.string,
      isLoading: PropTypes.bool,
      itemIdField: PropTypes.string,
    }),
    disabled: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
    onStateChange: PropTypes.func,
  };

  static defaultProps = {
    value: [],
    state: {
      items: [],
      placeholder: 'Search...',
      searchQuery: '',
      isLoading: false,
    },
    disabled: false,
  };

  render() {
    // for value that is array of object we use itemIdField, cuz internal select should get value as string - 
    // it has problem getting objects as value since it cant compare between them
    const mapFunc = (item) => {
      return {
        label: item.label, 
        value: this.props.state.itemIdField ? item.value[this.props.state.itemIdField] : item.value, 
        orgValue: item.value, 
      };
    };

    const selected = (this.props.value || []).map(mapFunc);
    const options = (this.props.state.items || []).map(mapFunc);

    return (
      <SelectInternal
        isMulti={true}
        value={selected} 
        onChange={this.onValueChange}
        options={options}
        placeholder={this.props.state.placeholder}
        isDisabled={this.props.disabled}
        isSearchable={true}
        isLoading={!!this.props.state.isLoading}
        inputValue={this.props.state.searchQuery}
        onInputChange={this.onSearchChange}
        styles={customStyles}
      />
    );
  }

  onValueChange = (selected) => {
    let newValue;
    newValue = !selected ? undefined : selected.map(item => { return { label: item.label, value: item.orgValue }; });
    this.props.onValueChange(newValue);
  }

  onSearchChange = (query) => {
    const newState = Object.assign({}, this.props.state);
    newState.searchQuery = query;
    this.props.onStateChange(newState);
  }
}
