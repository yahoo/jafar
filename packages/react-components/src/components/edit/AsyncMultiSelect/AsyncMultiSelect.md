<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const mockServerItems = [{
  id: '1',
  name: 'Ross Geller'
}, {
  id: '2',
  name: 'Monica Geller'
}, {
  id: '3',
  name: 'Rachel Green'
}, {
  id: '4',
  name: 'Chandler Bing'
}, {
  id: '5',
  name: 'Joey Tribbiani'
}, {
  id: '6',
  name: 'Phoebe Buffay'
}];

const mockService = {
  getByIds: (ids) => {
    return new Promise(resolve => {
      const item = mockServerItems.filter(x => ids.find(id => id === x.id));
      resolve(item);
    });
  },
  search: (query) => {
    return new Promise(resolve => {
      const items = mockServerItems.filter(x => x.name.toLowerCase().indexOf(query.toLowerCase()) > -1);
      resolve(items);
    });
  },
};

const model = {
  id: 'simple',
  fields: {
    bestFriends: {
      label: 'Best Friends',
      path: 'bestFriends',
      component: {
        name: 'AsyncMultiSelect',
        stateChange: { name: 'friendsStateChange' },
      },
      formatter: { name: 'friendsFormatter' },
      parser: { name: 'friendsParser' },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
    bestFriends: ['2', '3'], // remove to see initial value undefined, and required error
  },
};

const resources = {
  components: { 
    AsyncMultiSelect: {
      renderer: AsyncMultiSelect,
      stateChange: (props) => {    
        // if search filter has changed - return object with cleared items and 
        // change isLoading indication to true
        if (props.prevState && (props.prevState.searchQuery !== props.state.searchQuery)) {
          return { ...props.state, items: [], isLoading: true };
        }

        // else, if isLoading changes to true,
        // begin search and return promise which resolves an object of: { items: [ ... ], isLoading: false }
        if (props.state.isLoading) {
          return new Promise((resolve) => {
            // mock server call
            mockService.search(props.state.searchQuery).then((items) => {
              const viewItems = items.map(x => { return { label: x.name, value: x.id }; });
              resolve({ ...props.state, items: viewItems, isLoading: false }); // items should contain objects that has { label, value }
            });
          });
        }

        // else - no more changes needed
        return undefined;
      },
    },
  },
  conversions: {
    friendsFormatter: {
      func: (props) => {
        return new Promise((resolve) => {
          // mock server call
          mockService.getByIds(props.value).then((items) => {
            const viewItems = items.map(item => { return { label: item.name, value: item.id  }; });
            resolve(viewItems);
          });
        });  
      },
    },
    friendsParser: {
      func: (props) => {
        return props.value.map(item => item.value); // (props.value is the array [{ label, value }])
      },
    },
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='bestFriends' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
  value: [{ value: '3', label: 'Rachel Green' }],
};

<AsyncMultiSelect
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
  onStateChange={(newState) => {
    newState.items = [];
    newState.isLoading = true;
    setState({ state: newState });

    setTimeout(() => {
      const items = [{
        value: '1',
        label: 'Ross Geller'
      }, {
          value: '2',
          label: 'Monica Geller'
      }, {
          value: '3',
          label: 'Rachel Green'
      }, {
          value: '4',
          label: 'Chandler Bing'
      }, {
          value: '5',
          label: 'Joey Tribbiani'
      }, {
          value: '6',
          label: 'Phoebe Buffay'
      }];

      const filteredItems = items.filter(x => x.label.toLowerCase().indexOf(newState.searchQuery.toLowerCase()) > -1);
      const newState2 = Object.assign({}, newState);
      newState2.items = filteredItems;
      newState2.isLoading = false;
      setState({ state: newState2 });
    }, 700);
  }}
/>
```

  <h4>Value as object</h4>

```javascript
initialState = { 
  value: [{ value: { id: '3', first: 'Rachel', last: 'Green' }, label: 'Rachel Green' }],
  state: {
    itemIdField: 'id',
  }
};

<AsyncMultiSelect
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
  onStateChange={(newState) => {
    newState.items = [];
    newState.isLoading = true;
    setState({ state: newState });

    setTimeout(() => {
      const items = [{
        value: { id: '1', first: 'Ross', last: 'Geller' },
        label: 'Ross Geller'
      }, {
          value: { id: '2', first: 'Monica', last: 'Geller' },
          label: 'Monica Geller'
      }, {
          value: { id: '3', first: 'Rachel', last: 'Green' },
          label: 'Rachel Green'
      }, {
          value: { id: '4', first: 'Chandler', last: 'Bing' },
          label: 'Chandler Bing'
      }, {
          value: { id: '5', first: 'Joey', last: 'Tribbiani' },
          label: 'Joey Tribbiani'
      }, {
          value: { id: '6', first: 'Phoebe', last: 'Buffay' },
          label: 'Phoebe Buffay'
      }];

      const filteredItems = items.filter(x => x.label.toLowerCase().indexOf(newState.searchQuery.toLowerCase()) > -1);
      const newState2 = Object.assign({}, newState);
      newState2.items = filteredItems;
      newState2.isLoading = false;
      setState({ state: newState2 });
    }, 700);
  }}
/>
```

<h4>Disabled</h4>

```javascript
initialState = { 
  value: [{ value: '3', label: 'Rachel Green' }],
};

<AsyncMultiSelect
  value={state.value}
  state={state.state}
  disabled={true}
  onValueChange={(value) => {
    setState({ value });
  }}
  onStateChange={(newState) => {
    newState.items = [];
    newState.isLoading = true;
    setState({ state: newState });

    setTimeout(() => {
      const items = [{
        value: '1',
        label: 'Ross Geller'
      }, {
          value: '2',
          label: 'Monica Geller'
      }, {
          value: '3',
          label: 'Rachel Green'
      }, {
          value: '4',
          label: 'Chandler Bing'
      }, {
          value: '5',
          label: 'Joey Tribbiani'
      }, {
          value: '6',
          label: 'Phoebe Buffay'
      }];

      const filteredItems = items.filter(x => x.label.toLowerCase().indexOf(newState.searchQuery.toLowerCase()) > -1);
      const newState2 = Object.assign({}, newState);
      newState2.items = filteredItems;
      newState2.isLoading = false;
      setState({ state: newState2 });
    }, 700);
  }}
/>
```

<h4>Value undefined</h4>

```javascript
initialState = { 
  value: undefined,
};

<AsyncMultiSelect
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
  onStateChange={(newState) => {
    newState.items = [];
    newState.isLoading = true;
    setState({ state: newState });

    setTimeout(() => {
      const items = [{
        value: '1',
        label: 'Ross Geller'
      }, {
          value: '2',
          label: 'Monica Geller'
      }, {
          value: '3',
          label: 'Rachel Green'
      }, {
          value: '4',
          label: 'Chandler Bing'
      }, {
          value: '5',
          label: 'Joey Tribbiani'
      }, {
          value: '6',
          label: 'Phoebe Buffay'
      }];

      const filteredItems = items.filter(x => x.label.toLowerCase().indexOf(newState.searchQuery.toLowerCase()) > -1);
      const newState2 = Object.assign({}, newState);
      newState2.items = filteredItems;
      newState2.isLoading = false;
      setState({ state: newState2 });
    }, 700);
  }}
/>
```