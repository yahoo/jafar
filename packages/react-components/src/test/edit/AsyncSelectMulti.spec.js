/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { mount, shallow } from 'enzyme';
import AsyncMultiSelect from '../../edit/AsyncMultiSelect/index';

describe('<AsyncMultiSelect />', () => {
  let component;
  let onValueChangeSpy;
  let onStateChangeSpy;
  let state;
  let value;
  let disabled;
  let required;

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
    onStateChangeSpy = jest.fn();
    value = [{ value: 'ocean', label: 'Ocean' }];
    disabled = false;
    required = false;
    state = { 
      options:[
        { value: 'ocean', label: 'Ocean' },
        { value: 'blue', label: 'Blue' },
        { value: 'purple', label: 'Purple' },
      ],
    };
  });

  it('Should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('Should render provided data - disabled', () => {
    disabled = true;
    component = shallow(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('Should render provided data - required', () => {
    required = true;
    component = shallow(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  it('onValueChange - new value defined - ok', () => {
    component = mount(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    const controller = component.instance();
    const newValue = [{ value: 'purple', orgValue: 'purple', label: 'Purple' }];
    const expectedValue = [{ value: 'purple', label: 'Purple' }];
    controller.onValueChange(newValue);
    expect(onValueChangeSpy).toHaveBeenCalledWith(expectedValue);
  });

  it('onValueChange - new value undefined - ok', () => {
    component = mount(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    const controller = component.instance();
    const newValue = undefined;
    controller.onValueChange(newValue);
    expect(onValueChangeSpy).toHaveBeenCalledWith(undefined);
  });

  it('onSearchChange - ok', () => {
    component = mount(
      getComponent(value, state, disabled, required, onValueChangeSpy, onStateChangeSpy)
    );
    const controller = component.instance();
    const query = 'pu';
    controller.onSearchChange(query);
    const expectedState = state;
    expectedState.searchQuery = query;
    expect(onStateChangeSpy).toHaveBeenCalledWith(expectedState);
  });

  function getComponent(value, state, disabled, required, onValueChange, onStateChange) {
    return (
      <AsyncMultiSelect
        value={value}
        state={state}
        disabled={disabled}
        required={required}
        onValueChange={onValueChange}
        onStateChange={onStateChange}
      />
    );
  }
});
