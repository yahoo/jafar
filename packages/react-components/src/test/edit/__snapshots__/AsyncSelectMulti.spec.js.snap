// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AsyncMultiSelect /> Should render provided data - disabled 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={true}
  isLoading={false}
  isMulti={true}
  isSearchable={true}
  onChange={[Function]}
  onInputChange={[Function]}
  options={Array []}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Array [
      Object {
        "label": "Ocean",
        "orgValue": "ocean",
        "value": "ocean",
      },
    ]
  }
/>
`;

exports[`<AsyncMultiSelect /> Should render provided data - required 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={false}
  isLoading={false}
  isMulti={true}
  isSearchable={true}
  onChange={[Function]}
  onInputChange={[Function]}
  options={Array []}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Array [
      Object {
        "label": "Ocean",
        "orgValue": "ocean",
        "value": "ocean",
      },
    ]
  }
/>
`;

exports[`<AsyncMultiSelect /> Should render provided data 1`] = `
<StateManager
  defaultInputValue=""
  defaultMenuIsOpen={false}
  defaultValue={null}
  isDisabled={false}
  isLoading={false}
  isMulti={true}
  isSearchable={true}
  onChange={[Function]}
  onInputChange={[Function]}
  options={Array []}
  styles={
    Object {
      "clearIndicator": [Function],
      "control": [Function],
      "dropdownIndicator": [Function],
      "menu": [Function],
      "menuList": [Function],
      "singleValue": [Function],
      "valueContainer": [Function],
    }
  }
  value={
    Array [
      Object {
        "label": "Ocean",
        "orgValue": "ocean",
        "value": "ocean",
      },
    ]
  }
/>
`;
